File: codejam/python/2009/1B/1B_SquareMath.py
import sys
from collections import deque, namedtuple

## Input crap
infile = sys.stdin
intokens = deque()
def getTokens(): 
    while not intokens:
        for c in infile.readline().rstrip().split() :
            intokens.append(c)    
def gs(): getTokens(); return intokens.popleft()
def gi(): return int(gs())
def gf(): return float(gs())
def gbs(): return [c for c in gs()]
def gis(n): return [gi() for i in range(n)]
def ia(m): return [0] * m
def iai(m,v): return [v] * m
def twodi(n,m,v): return [iai(m,v) for i in range(n)]
def fill2(m) : r = gis(2*m); return r[0::2],r[1::2]
def fill3(m) : r = gis(3*m); return r[0::3],r[1::3],r[2::3]
def fill4(m) : r = gis(4*m); return r[0::4],r[1::4],r[2::4],r[3::4]
MOD = 998244353
##MOD = 1000000007

def main() :
    if len(sys.argv) > 1 : global infile; infile = open(sys.argv[1],'rt')
    ## PROGRAM STARTS HERE
    T = gi()
    for tt in range(1,T+1) :
        W,Q = gi(),gi(); bd = [gs() for _ in range(W)]; QQ = [gi() for _ in range(Q)]
        midx = namedtuple("midx","i j v")
        mval = namedtuple("mval","l s")
        ansarr = [ mval(0,"") for _ in range(251) ]
        numleft = 0
        for q in QQ :
            if ansarr[q].l == 0 :
                numleft += 1; ansarr[q] = mval(-1,"")
        que = deque(); rnd = 1; lkup = {}
        for i in range(W) :
            for j in range(W) :
                if bd[i][j] in "-+" : continue
                v = int(bd[i][j])
                if ansarr[v].l == -1 : numleft -= 1; ansarr[v] = mval(1,bd[i][j])
                lkup[midx(i,j,v)] = mval(1,bd[i][j])
                que.append(midx(i,j,v))
        deltas = [(0,-1),(0,1),(-1,0),(1,0)]
        while que :
            m = que.popleft()
            mv = lkup[m]
            if rnd == mv.l :
                if numleft == 0 : break
                rnd += 1
            for (d1,d2) in deltas :
                i1,j1 = m.i+d1,m.j+d2
                if i1 < 0 or i1 >= W or j1 < 0 or j1 >= W : continue
                sgn = 1 if bd[i1][j1] == '+' else -1
                for (d3,d4) in deltas :
                    i2,j2 = i1+d3,j1+d4
                    if i2 < 0 or i2 >= W or j2 < 0 or j2 >= W : continue
                    v = m.v+sgn*int(bd[i2][j2])
                    key = midx(i2,j2,v)
                    val = mval(1000000,"")
                    if key in lkup : val = lkup[key]
                    if val.l < rnd : continue
                    val2 = mval(rnd,mv.s + (bd[i1][j1] + bd[i2][j2]))
                    if val.l == val2.l and val.s <= val2.s : continue
                    if key not in lkup : que.append(key)
                    lkup[key] = val2
                    if key.v >= 0 and key.v <= 250 and (ansarr[key.v].l == -1 or ansarr[key.v].l == rnd and val2.s < ansarr[key.v].s) :
                        if (ansarr[key.v].l == -1) : numleft -= 1
                        ansarr[key.v] = val2
        print(f"Case #{tt}:")
        for q in QQ :
            print(ansarr[q].s)

if __name__ == "__main__" :
    main()

File: codejam/cpp/2009/1B/1B_SquareMath.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pi;
#define FOR(i,a) for (ll i = 0; i < (a); i++)
#define len(x) (ll) x.size()
const ll INF = 1LL << 62;
const ll MOD = 1000000007;
//const ll MOD = 998244353;
const double PI = 4*atan(double(1.0));
vector<vector<ll>> twodi(ll N, ll M, ll v) { vector<vector<ll>> res(N); for (auto &vv : res) vv.resize(M,v); return res; }

struct midx { ll i,j,v; };
bool operator < (const midx a, const midx b) { return a.i < b.i || a.i == b.i && (a.j < b.j || a.j == b.j && (a.v < b.v)); }

struct mval {ll l; string s;};
int main (int argc, char **argv) {
    if (argc > 1) { freopen(argv[1],"r",stdin); }
    ios_base::sync_with_stdio(false);cin.tie(0);
    // PROGRAM STARTS HERE
    ll T; cin >> T;
    for (ll tt=1;tt<=T;tt++) {
        ll W,Q; cin >> W >> Q;
        vector<string> bd(W); FOR(i,W) { cin >> bd[i]; }
        vi QQ(Q); FOR(i,Q) { cin >> QQ[i]; }
        vector<mval> ansarr(251); FOR(i,251) ansarr[i] = {0,""};
        ll numleft = 0; for (auto &q : QQ) if (ansarr[q].l == 0) { ansarr[q].l = -1; numleft++; }
        map<midx,mval> lkup;
        vector<midx> que,nque;
        FOR(i,W) {
            FOR(j,W) {
                if (bd[i][j] != '+' && bd[i][j] != '-') { 
                    ll v = ll(bd[i][j]-'0');
                    midx m = {i,j,v};
                    mval vv = {1,bd[i].substr(j,1)};
                    if (v >= 0 && v <= 250 && ansarr[v].l == -1) { numleft--; ansarr[v] = vv; } 
                    lkup[m] = {1,bd[i].substr(j,1)};
                    que.push_back(m);
                }
            }
        }
        vector<pair<ll,ll>> deltas{{-1,0},{1,0},{0,-1},{0,1}};
        ll rnd = 1;
        while (numleft > 0) {
            rnd++;
            for (auto &m : que) {
                for (auto &p1 : deltas) {
                    auto i1 = m.i+p1.first; auto j1 = m.j+p1.second;
                    if (i1 < 0 || j1 < 0 || i1 >= W || j1 >= W) continue;
                    ll sgn = bd[i1][j1] == '+' ? 1 : -1;
                    for (auto &p2 : deltas) {
                        auto i2 = i1+p2.first; auto j2 = j1+p2.second;
                        if (i2 < 0 || j2 < 0 || i2 >= W || j2 >= W) continue;
                        ll v = bd[i2][j2]-'0';
                        midx m2({i2,j2,m.v+sgn*v});
                        auto it = lkup.find(m2);
                        if (it != lkup.end() && it->second.l < rnd) continue;
                        string path(lkup[m].s); path.push_back(bd[i1][j1]); path.push_back(bd[i2][j2]);
                        if (it != lkup.end() && it->second.l == rnd && path >= it->second.s) continue;
                        if (it == lkup.end()) { nque.push_back(m2); }
                        mval v2({rnd,path});
                        lkup[m2] = v2;
                        if (m2.v >= 0 && m2.v <= 250 && (ansarr[m2.v].l == -1 || ansarr[m2.v].l == rnd && path < ansarr[m2.v].s)) {
                            if (ansarr[m2.v].l == -1) numleft--;
                            ansarr[m2.v] = v2;
                        }
                    }
                }
            }
            swap(que,nque); nque.clear();
        }
        printf("Case #%lld:\n",tt);
        for (auto &qq : QQ) { printf("%s\n",ansarr[qq].s.c_str()); }
    }
}



